src/linear_algebra/matrix/transvection.lean

```diff
@@ -0,0 +1,754 @@
+    simp only [←matrix.mul_assoc, to_matrix_mk, list.prod_cons, mul_eq_mul, list.map],
+    rw [matrix.mul_assoc (L.map to_matrix).prod],
+    exact hLL' }
+end
+
+/-- Inductive step for the reduction: if one knows that any size `r` matrix can be reduced to
+diagonal form by elementary operations, then one deduces it for matrices over `fin r ⊕ unit`. -/
+lemma exists_list_transvec_mul_mul_list_transvec_eq_diagonal_induction
+  (IH : ∀ (M : matrix (fin r) (fin r) 𝕜),
+    ∃ (L₀ L₀' : list (transvection_struct (fin r) 𝕜)) (D₀ : (fin r) → 𝕜),
+      (L₀.map to_matrix).prod ⬝ M ⬝ (L₀'.map to_matrix).prod = diagonal D₀)
+  (M : matrix (fin r ⊕ unit) (fin r ⊕ unit) 𝕜) :
+  ∃ (L L' : list (transvection_struct (fin r ⊕ unit) 𝕜)) (D : fin r ⊕ unit → 𝕜),
+    (L.map to_matrix).prod ⬝ M ⬝ (L'.map to_matrix).prod = diagonal D :=
+begin
+  rcases exists_is_two_block_diagonal_list_transvec_mul_mul_list_transvec M with ⟨L₁, L₁', hM⟩,
+  let M' := (L₁.map to_matrix).prod ⬝ M ⬝ (L₁'.map to_matrix).prod,
+  let M'' := to_blocks₁₁ M',
+  rcases IH M'' with ⟨L₀, L₀', D₀, h₀⟩,
+  set c := M' (inr star) (inr star) with hc,
+  refine ⟨L₀.map (sum_inl unit) ++ L₁, L₁' ++ L₀'.map (sum_inl unit),
+    sum.elim D₀ (λ _, M' (inr star) (inr star)), _⟩,
+  suffices :
+    (L₀.map (to_matrix ∘ sum_inl unit)).prod ⬝ M' ⬝ (L₀'.map (to_matrix ∘ sum_inl unit)).prod =
+      diagonal (sum.elim D₀ (λ _, c)), by simpa [M', matrix.mul_assoc, c],
+  have : M' = from_blocks M'' 0 0 (diagonal (λ _, c)),
+  { rw ← from_blocks_to_blocks M',
+    congr,
+    { exact hM.1 },
+    { exact hM.2 },
+    { ext i j,  rw [hc, to_blocks₂₂], congr } },
+  rw this,
+  simp [h₀],
+end
+
+variables {n p} [fintype n] [fintype p]
+
+/-- Reduction to diagonal form by elementary operations is invariant under reindexing. -/
+lemma reindex_exists_list_transvec_mul_mul_list_transvec_eq_diagonal (M : matrix p p 𝕜) (e : p ≃ n)
+  (H : ∃ (L L' : list (transvection_struct n 𝕜)) (D : n → 𝕜),
+    (L.map to_matrix).prod ⬝ (matrix.reindex_alg_equiv 𝕜 e M) ⬝ (L'.map to_matrix).prod
+      = diagonal D) :
+  ∃ (L L' : list (transvection_struct p 𝕜)) (D : p → 𝕜),
+    (L.map to_matrix).prod ⬝ M ⬝ (L'.map to_matrix).prod = diagonal D :=
+begin
+  rcases H with ⟨L₀, L₀', D₀, h₀⟩,
+  refine ⟨L₀.map (reindex_equiv e.symm), L₀'.map (reindex_equiv e.symm), D₀ ∘ e, _⟩,
+  have : M = reindex_alg_equiv 𝕜 e.symm (reindex_alg_equiv 𝕜 e M),
+    by simp only [equiv.symm_symm, minor_minor, reindex_apply, minor_id_id, equiv.symm_comp_self,
+      reindex_alg_equiv_apply],
+  rw this,
+  simp only [to_matrix_reindex_equiv_prod, list.map_map, reindex_alg_equiv_apply],
+  simp only [← reindex_alg_equiv_apply, ← reindex_alg_equiv_mul, h₀],
+  simp only [equiv.symm_symm, reindex_apply, minor_diagonal_equiv, reindex_alg_equiv_apply],
+end
+
+/-- Any matrix can be reduced to diagonal form by elementary operations. Formulated here on `Type`
+because we will make an induction using `fin r`.
```

Comment:
```suggestion
/-- Any matrix can be reduced to diagonal form by elementary operations. Formulated here on `Type 0`
because we will make an induction using `fin r`.
```
(It took me a moment to understand that it's the universe restriction that is relevant here, so this change would have been helpful to me.)